kubetest/dumplogs.go

```diff
@@ -0,0 +1,261 @@
+}
+
+// newLogDumper is the constructor for a logDumper
+func newLogDumper(sshConfig *ssh.ClientConfig, artifactsDir string) (*logDumper, error) {
+	d := &logDumper{
+		sshConfig:    sshConfig,
+		artifactsDir: artifactsDir,
+	}
+
+	d.services = []string{
+		"node-problem-detector",
+		"kubelet",
+		"docker",
+		"kops-configuration",
+		"protokube",
+	}
+	d.files = []string{
+		"kube-apiserver",
+		"kube-scheduler",
+		"rescheduler",
+		"kube-controller-manager",
+		"etcd",
+		"etcd-events",
+		"glbc",
+		"cluster-autoscaler",
+		"kube-addon-manager",
+		"fluentd",
+		"kube-proxy",
+		"node-problem-detector",
+		"cloud-init-output",
+		"startupscript",
+		"kern",
+		"docker",
+	}
+
+	return d, nil
+}
+
+// DumpAllNodes connects to every node and dumps the logs
+func (d *logDumper) DumpAllNodes() error {
+	nodes, err := kubectlGetNodes()
+	if err != nil {
+		return err
+	}
+
+	for i := range nodes.Items {
+		node := &nodes.Items[i]
+
+		host := ""
+		for _, address := range node.Status.Addresses {
+			if address.Type == "ExternalIP" {
+				host = address.Address
+				break
+			}
+		}
+
+		if host == "" {
+			log.Printf("could not find address for %v", node.Metadata.Name)
+			continue
+		}
+
+		log.Printf("Dumping node %s", node.Metadata.Name)
+
+		d, err := d.connectToNode(node.Metadata.Name, host)
+		if err != nil {
+			log.Printf("could not connect to %s (%s): %v", node.Metadata.Name, host, err)
+			continue
+		}
+
+		errors := d.dump()
+		for _, e := range errors {
+			log.Printf("error dumping %s: %v", node.Metadata.Name, e)
+		}
+
+		if err := d.Close(); err != nil {
+			log.Printf("error closing connection to %s: %v", node.Metadata.Name, err)
+		}
+	}
+
+	return nil
+}
+
+// logDumperNode holds state for a particular node we are dumping
+type logDumperNode struct {
+	client *ssh.Client
+	dumper *logDumper
+
+	dir string
+}
+
+// connectToNode makes an SSH connection to the node and returns a logDumperNode
+func (d *logDumper) connectToNode(nodeName string, host string) (*logDumperNode, error) {
+	client, err := ssh.Dial("tcp", host+":22", d.sshConfig)
+	if err != nil {
+		return nil, fmt.Errorf("unable to SSH to %q: %v", host, err)
+	}
+	return &logDumperNode{
+		client: client,
+		dir:    filepath.Join(d.artifactsDir, nodeName),
+		dumper: d,
+	}, nil
+}
+
+// logDumperNode cleans up any state in the logDumperNode
+func (d *logDumperNode) Close() error {
+	return d.client.Close()
+}
+
+// dump captures the well-known set of logs
+func (d *logDumperNode) dump() []error {
+	var errors []error
+
+	// Capture kernel log
+	if err := d.shellToFile("sudo journalctl --output=short-precise -k", filepath.Join(d.dir, "kern.log")); err != nil {
+		errors = append(errors, err)
+	}
+
+	// Capture logs from any systemd services in our list, that are registered
+	services, err := d.listSystemdUnits()
+	if err != nil {
+		errors = append(errors, fmt.Errorf("error listing systemd services: %v", err))
+	}
+	for _, s := range d.dumper.services {
```

Comment:
Yes, but I figured I could do this later.  This PR is already way too big :-)

I actually think it's probably better to run the nodes in parallel, rather than the jobs against a particular node, but no strong opinion...